File: bag/test_bag.py
import bag
import unittest

class TestBag(unittest.TestCase):
    def setUp(self):
        self.b = bag.Bag()

    def tearDown(self):
        del self.b

    def test_num_items(self):
        self.assertEqual(self.b.num_items(), 0)
        self.b.add("summer")
        self.assertEqual(self.b.num_items(), 1)
        x = self.b.grab() # grabbing does not remove
        self.assertEqual(self.b.num_items(), 1)
        self.b.add("winter")
        self.assertEqual(self.b.num_items(), 2)
        self.b.remove("summer")
        self.assertEqual(self.b.num_items(), 1)
        
    def test_add(self):
        # when you have only one element, grabbing should return it
        self.b.add("summer")
        self.assertTrue(self.b.grab() == "summer")
        self.b.remove("summer")
        self.b.add(555)
        self.assertTrue(self.b.grab() == 555)

    def test_remove(self):
        # when you remove the only element there should be none left
        self.b.add("summer")
        self.b.remove("summer")
        self.assertEqual(self.b.num_items(),0)

    def test_contains(self):
        self.assertFalse(self.b.contains("summer"))
        self.b.add("summer")
        self.assertTrue(self.b.contains("summer"))

    def test_grab(self):
        # with 3 items in the bag, enough grabs ought to return different items
        # the number of items should never change
        self.b.add("summer")
        self.b.add("winter")
        self.b.add(555)
        self.assertEqual(self.b.num_items(), 3)
        grabbed = []
        for i in range(50):
            x = self.b.grab()
            if x not in grabbed:
                grabbed.append(x)

        self.assertTrue(len(grabbed)) > 1          
        self.assertEqual(self.b.num_items(), 3)

    def test_to_array(self):
        # the length of returned array should match num_items
        # something grabbed should be in the array
        self.b.add("summer")
        self.b.add("winter")
        self.b.add(555)
        arr = self.b.to_array()
        self.assertEqual(self.b.num_items(), len(arr))
        self.assertTrue(self.b.grab() in arr)


if __name__ == '__main__':
    unittest.main()

File: bag/test_bag.rb
# run some tests on Bag class

require "Bag"
require "test/unit"

class TestBag < Test::Unit::TestCase

    def setup
        @b = Bag.new
    end

    def test_num_items
        assert_equal(0, @b.num_items)
        @b.add("summer")
        assert_equal(1, @b.num_items)
        @b.remove("summer")
        assert_equal(0, @b.num_items)
    end

    def test_add
        @b.add("summer")
        @b.add("winter")
        @b.add(5)
        assert(@b.contains("summer"))
        assert(@b.contains("winter"))
        assert(@b.contains(5))
        assert_equal(3, @b.num_items)
    end

    def test_remove
        @b.add("summer")
        @b.add("winter")
        @b.add(5)
        assert(@b.contains("winter"))
        @b.remove("winter")
        assert(!@b.contains("winter"))
        assert_equal(2,@b.num_items)
        @b.remove("blue")
        assert_equal(2,@b.num_items)
    end

    def test_grab
        @b.add("summer")
        @b.add("winter")
        @b.add(5)
        grabbed = Array.new
        50.times {g = @b.grab; grabbed.push(g) if grabbed.index(g) == nil}
        assert(grabbed.length > 1)
    end

    def test_contains
        @b.add("summer")
        assert(@b.contains("summer"))
        assert(!@b.contains("blue"))
    end

    def test_to_array
        @b.add("summer")
        @b.add("winter")
        @b.add(5)
        arr = @b.to_array
        assert_not_nil(arr.index("summer"))
        assert_not_nil(arr.index("winter"))
        assert_not_nil(arr.index(5))
        assert_equal(3, arr.length)
    end

    def teardown
        # nothing
    end
end
